crates/typst/src/layout/flow/distribute.rs

```diff
@@ -0,0 +1,510 @@
+use super::{
+    Child, Composer, FlowResult, LineChild, MultiChild, MultiSpill, PlacedChild,
+    SingleChild, Stop, Work,
+};
+use crate::introspection::Tag;
+use crate::layout::{
+    Abs, Axes, FixedAlignment, Fr, Frame, FrameItem, Point, Region, Regions, Rel, Size,
+};
+use crate::utils::Numeric;
+
+/// Distributes as many children as fit from `composer.work` into the first
+/// region and returns the resulting frame.
+pub fn distribute(composer: &mut Composer, regions: Regions) -> FlowResult<Frame> {
+    let mut distributor = Distributor {
+        composer,
+        regions,
+        items: vec![],
+        sticky: None,
+        stickable: false,
+    };
+    let init = distributor.snapshot();
+    let forced = match distributor.run() {
+        Ok(()) => true,
+        Err(Stop::Finish(forced)) => forced,
+        Err(err) => return Err(err),
+    };
+    let region = Region::new(regions.size, regions.expand);
+    distributor.finalize(region, init, forced)
+}
+
+/// State for distribution.
+struct Distributor<'a, 'b, 'c, 'd, 'e> {
```

Comment:
With the number of lifetime parameters here, I think it’d be helpful to give them more meaningful names.

Reply:
Sort of unfortunate that so many are needed. It's because of variance. The only interesting ones are 'a and 'b. These are used across the flow layout.

Reply:
Could you possibly name just them and leave the other three as generic `'a`, `'b`, `'c`? This would make the distinction between the important

Reply:
I'm not a huge fan of lifetimes with long names because I feel like they clutter things without adding that much value. The difference between 'a and 'b is not that interesting and if Rust would've let me, I would've used 'a for both. What matters in my opinion is that 'a and 'b mean the same thing across their usages in one subsystem. 

But I agree that it's relevant to differentiate between the two shared ones and the three irrelevant ones. Maybe I could just do what I did in realize.rs and call them 'a, 'b, 'x, 'y, 'z and then add a doc comment explaining things.